adr/ADR-31.md

```diff
@@ -0,0 +1,182 @@
+assured that a subsequent call to the Get API will return that message as the read will go a server that defines 
+_most current_. 
+
+In contrast, _Direct Get_ does not assure read-after-write coherency as responders may be non-leader stream servers 
+(that may not have yet applied the latest consensus writes) or MIRROR downstream servers that have not yet _consumed_ 
+the latest consensus writes from upstream.
+
+## Implementation
+
+### Stream property: Allow Direct 
+
+- Stream configuration adds a new Allow Direct boolean property: `allow_direct`
+- Allow Direct is always set to `true` by the server when maximum messages per subject, `max_msgs_per_subject`, is configured to be > 0 (a max limit is specified)
+- If the user passes Allow Direct explicitly in stream create or edit request, the value will be overriden
+based on `max_msgs_per_subject`
+
+> Allow Direct is set automatically based on the inferred use case of the stream. Maximum messages per subject is a
+tell-tale of a stream that is a KV bucket.
+
+### Direct Get API 
+
+When Allow Direct is true, each of the stream's servers configures a responder and subscribes to 
+`$JS.API.DIRECT.GET.<stream>` with fixed queue group `_sys_`. 
+
+> Note: If Allow Direct is false, there will be no responder at the Direct Get API for a stream. Clients that make
+> requests will receive no reply message and will time out.
+
+###### Request
+
+Clients may make requests with the same payload as the Get message API populating the following server struct:
+
+ ```text
+Seq     uint64 `json:"seq,omitempty"`
+LastFor string `json:"last_by_subj,omitempty"`
+NextFor string `json:"next_by_subj,omitempty"`
+```
+
+Example request payloads:
+* `{last_by_subj: string}` - get the last message having the subject
+* `{seq: number}` - get a message by sequence
+* `{next_by_subj: string}` - get the first message (lowest seq) having the specified subject
+* `{seq: number, next_by_subj: string}` - get the first message with a seq >= to the input seq that has the specified subject
+
+#### Subject-Appended Direct Get API
+
+The purpose of this form is so that environmnents may choose to apply subject-based interest restrictions (user permissions
+within an account and/or cross-account export/import grants) such that only specific
+subjects in stream may be read (vs any subject in the stream).
+
+When Allow Direct is true, each of the stream's servers will also subscribe to `$JS.API.DIRECT.GET.<stream>.>` with
+fixed queue group `_sys_`.  Requests to this API endpoint will be interpreted as a shortcut for `last_by_subj` request
+where `last_by_subj` is derived by the token (or series of tokens) following the stream name rather than request payload.
+
+It is an error (408) if a client calls Subject-Appended Direct Get and includes a request payload.
+
+#### Response Format 
+
+Direct Get may return an error code:
```

Comment:
Just maybe add that it is through the use of headers, not a typical JetStream response object.

Reply:
Yes, I show in an example but good suggestion to clarify here.  Will add.